### PR TITLE
Add maxPage config key to limit pagination results as a whole.

### PR DESCRIPTION
**Pull request by dereuromark**
Resolves https://github.com/cakephp/cakephp/issues/9446

It happened quite a few times now that people were looking for a way to limit results of pagination similar as results would be limited non-paginated.
The limit option for paginations, though, is used per page. There is nothing right now to limit the results in general here.

In order to allow only x records total to be fetched via pagination, one would need to configure `maxLimit` and `maxPage` accordingly as outlined below.

The helper will also only display those pages as it gets its data from the component, which in case of such a maxPage overflow will adjust the count and page number.

@aavrug Can you confirm that this solves your issue?

### CODE REVIEW

File: src/Controller/Component/PaginatorComponent.php [hunk 1]
```diff
@@ -418,4 +423,23 @@ public function checkLimit(array $options)
...
+    /**
+     * Check the page parameter and ensure it's within the maxPage bounds.
+     *
+     * @param array $options An array of options with a page key to be checked.
+     * @return array An array of options for pagination
+     */
+    public function checkPage(array $options)
+    {
+        $options['page'] = (int)$options['page'];
+        if (empty($options['page']) || $options['page'] < 1) {
```
Comment (chinpei215) [inline src/Controller/Component/PaginatorComponent.php]: > empty($options['page']) || 

Seems redundant.

Comment (dereuromark) [inline src/Controller/Component/PaginatorComponent.php]: You are right

File: src/Controller/Component/PaginatorComponent.php [hunk 2]
```diff
@@ -188,6 +190,9 @@ public function paginate($object, array $settings = [])
         $results = $query->all();
         $numResults = count($results);
         $count = $numResults ? $query->count() : 0;
+        if ($options['maxPage'] && $count > $options['maxPage'] * $options['limit']) {
```
Comment (dereuromark) [inline src/Controller/Component/PaginatorComponent.php]: That should use maxlimit actually

Comment (markstory) [inline src/Controller/Component/PaginatorComponent.php]: Why? The current limit influences the number of pages. Not the maxLimit.

**Comment by markstory** [comment]
Doesn't the PaginatorHelper also need to be aware of maxPage? Right now it *could* generate links that would not be accessible.

**Comment by dereuromark** [comment]
Doesn't the helper rely on the data from component? Which then should be "fixed" according to the expectation.

File: src/Controller/Component/PaginatorComponent.php (hunk 2 above)
Comment (dereuromark) [inline src/Controller/Component/PaginatorComponent.php]: The maxLimit is actually the one restricting the "total limit" of records. limit is only for display per page.

**Comment by markstory** [comment]
@dereuromark Imagine a dataset of 1000 rows. For some reason you've chosen to set `maxLimit = 10` and `maxPage = 2`. Currently wouldn't PaginatorHelper::numbers() generate links for pages 3, 4, 5 when you are on page 1?

**Comment by dereuromark** [comment]
This really only makes sense if one does not change the sort direction in any way for any of the fields.
Otherwise the resulting records would be totally different all the time.
So in light of this it makes no sense to add after all.
Maybe we can improve docs here for what limit does?

**Comment by aavrug** [comment]
What is the expected behavior for this $users = $this->paginate($this->Users, ['maxLimit' => 2, 'maxPage' => 2]); ?

**Comment by ADmad** [comment]
@aavrug The PR was closed without merging so there is no `maxPage` option. Read the discussion on why it was a bad idea to add that option.

**Comment by aavrug** [comment]
@ADmad ok got it. So, again we are on same point and there is no such option ```like``` as it was in CakePHP-2.X.

**Comment by lorenzo** [comment]
@aavrug there was no similar option min cake 2

**Comment by josegonzalez** [comment]
There was never a `maxPage` option in 2.x. `limit` and `maxLimit` have always referred to the max number of records that can be retrieved in a single pagination call, not the number of records total that could be retrieved if you paginated through a given set of options.

A better way to implement this is to write a custom finder where you can validate/overwrite any passed in query options such as `page`. That would provide the needed flexibility without adding extra code to the cake core. I personally have never used a "max page" option, and while I can see it's usefulness in certain types of applications, I don't think it's a common use case and adding code complexity to the core gets a 👎 from me.

We might be better suited to rename the `maxLimit` option to `maxRecordLimit` or `maxPerPageLimit`, heh.

**Comment by dereuromark** [comment]
And Limit should be named perPageLimit

**Comment by aavrug** [comment]
@lorenzo I have used it in my older projects http://book.cakephp.org/2.0/en/core-libraries/components/pagination.html#limit-the-maximum-number-of-rows-that-can-be-fetched. There should be an option in CakePHP-3 because it is a common requirement that if I have 1000 records and I need to display only the first 25 records why I'll fetch all the records?

**Comment by dereuromark** [comment]
Oh geeze, someone really needs to fix up that documenation now.
Or this will never end.

**Comment by aavrug** [comment]
@dereuromark That means I again misunderstood the doc? :(

**Comment by josegonzalez** [comment]
@aavrug that `maxLimit` is functionally the same as the `maxLimit` in the current CakePHP 3.x release. 

In all versions of CakePHP: 

- The pagination request *does not* retrieve all the records at once, just the number set by your `limit` option.
- All things being equal, `limit` is the lower bound of records that will be retrieved for a single page, and `maxLimit` is the upper bound of records that will be retrieved for a single page.
- There is a single `count` call that is run against your pagination query that you might be confusing for the number of records retrieved. That is the total number of records that *could* be retrieved if you paginated through every page in your pagination result set.
- You cannot limit the max total number of records that would be paginated with any existing pagination option.
- You cannot limit the max number of pages in your pagination result set with any existing pagination option.

**Comment by aavrug** [comment]
@josegonzalez I think now I have a clear view and sorry all of you if I bothered you.

**Pull request closed without merging**